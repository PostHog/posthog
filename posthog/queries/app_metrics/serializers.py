from rest_framework import serializers


class AppMetricsRequestSerializer(serializers.Serializer):
    category = serializers.ChoiceField(
        # Keep in sync with plugin-server/src/worker/ingestion/app-metrics.ts
        choices=["processEvent", "onEvent", "exportEvents", "scheduledTask", "webhook"],
        help_text="What to gather metrics for",
        required=True,
    )
    date_from = serializers.CharField(
        default="-30d",
        help_text="What date to filter the results from. Can either be a date `2021-01-01`, or a relative date, like `-7d` for last seven days, `-1m` for last month, `mStart` for start of the month or `yStart` for the start of the year.",
    )
    date_to = serializers.CharField(
        required=False,
        help_text="What date to filter the results to. Can either be a date `2021-01-01`, or a relative date, like `-7d` for last seven days, `-1m` for last month, `mStart` for start of the month or `yStart` for the start of the year.",
    )
    job_id = serializers.CharField(help_text="Set this to filter results to a particular job", required=False)


class AppMetricsErrorsRequestSerializer(serializers.Serializer):
    category = serializers.ChoiceField(
        # Keep in sync with plugin-server/src/worker/ingestion/app-metrics.ts
        choices=["processEvent", "onEvent", "exportEvents", "scheduledTask", "webhook"],
        help_text="What to gather errors for",
        required=True,
    )
    error_type = serializers.CharField(required=True, help_text="What error type to filter for.")
    job_id = serializers.CharField(help_text="Set this to filter results to a particular job", required=False)
