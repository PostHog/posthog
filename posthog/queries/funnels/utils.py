from typing import Type

from posthog.constants import FunnelOrderType
from posthog.models.filters import Filter
from posthog.queries.funnels import ClickhouseFunnelBase


def get_funnel_order_class(filter: Filter) -> Type[ClickhouseFunnelBase]:
    from posthog.queries.funnels import ClickhouseFunnel, ClickhouseFunnelStrict, ClickhouseFunnelUnordered

    if filter.funnel_order_type == FunnelOrderType.UNORDERED:
        return ClickhouseFunnelUnordered
    elif filter.funnel_order_type == FunnelOrderType.STRICT:
        return ClickhouseFunnelStrict
    return ClickhouseFunnel


def get_funnel_order_actor_class(filter: Filter):
    from posthog.queries.funnels import (
        ClickhouseFunnelActors,
        ClickhouseFunnelStrictActors,
        ClickhouseFunnelUnorderedActors,
    )

    if filter.funnel_order_type == FunnelOrderType.UNORDERED:
        return ClickhouseFunnelUnorderedActors
    elif filter.funnel_order_type == FunnelOrderType.STRICT:
        return ClickhouseFunnelStrictActors
    return ClickhouseFunnelActors
