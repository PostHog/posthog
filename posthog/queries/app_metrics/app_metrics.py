import json
from datetime import datetime, timedelta

from django.utils.timezone import now

from posthog.client import sync_execute
from posthog.models.app_metrics.sql import (
    QUERY_APP_METRICS_DELIVERY_RATE,
    QUERY_APP_METRICS_ERROR_DETAILS,
    QUERY_APP_METRICS_ERRORS,
    QUERY_APP_METRICS_TIME_SERIES,
)
from posthog.models.event.util import format_clickhouse_timestamp
from posthog.models.filters.mixins.base import IntervalType
from posthog.models.team.team import Team
from posthog.queries.app_metrics.serializers import (
    AppMetricsErrorsRequestSerializer,
    AppMetricsRequestSerializer,
)
from posthog.queries.util import format_ch_timestamp, get_time_in_seconds_for_period
from posthog.utils import relative_date_parse


class TeamPluginsDeliveryRateQuery:
    QUERY = QUERY_APP_METRICS_DELIVERY_RATE

    def __init__(self, team: Team):
        self.team = team

    def run(self):
        results = sync_execute(
            self.QUERY,
            {
                "team_id": self.team.pk,
                "from_date": format_clickhouse_timestamp(datetime.now() - timedelta(hours=24)),
            },
        )
        return dict(results)


class AppMetricsQuery:
    QUERY = QUERY_APP_METRICS_TIME_SERIES

    def __init__(self, team: Team, plugin_config_id: int, filter: AppMetricsRequestSerializer):
        self.team = team
        self.plugin_config_id = plugin_config_id
        self.filter = filter

    def run(self):
        query, params = self.query()
        dates, successes, successes_on_retry, failures = sync_execute(query, params)[0]
        return {
            "dates": [
                timestamp.strftime("%Y-%m-%d{}".format(" %H:%M:%S" if self.interval == "hour" else ""))
                for timestamp in dates
            ],
            "successes": successes,
            "successes_on_retry": successes_on_retry,
            "failures": failures,
            "totals": {
                "successes": sum(successes),
                "successes_on_retry": sum(successes_on_retry),
                "failures": sum(failures),
            },
        }

    def query(self):
        job_id = self.filter.validated_data.get("job_id")
        query = self.QUERY.format(
            job_id_clause="AND job_id = %(job_id)s" if job_id is not None else "",
            interval_function=self.interval_function,
        )

        return query, {
            "team_id": self.team.pk,
            "plugin_config_id": self.plugin_config_id,
            "category": self.filter.validated_data.get("category"),
            "job_id": job_id,
            "date_from": format_ch_timestamp(self.date_from),
            "date_to": format_ch_timestamp(self.date_to),
            "timezone": self.team.timezone,
            "interval": self.interval,
            "with_fill_step": get_time_in_seconds_for_period(self.interval),
        }

    @property
    def date_from(self):
        return relative_date_parse(
            self.filter.validated_data.get("date_from"),
            self.team.timezone_info,
            always_truncate=True,
        )

    @property
    def date_to(self):
        date_to_string = self.filter.validated_data.get("date_to")
        return (
            relative_date_parse(date_to_string, self.team.timezone_info, always_truncate=True)
            if date_to_string is not None
            else now()
        )

    @property
    def interval(self) -> IntervalType:
        if self.date_to - self.date_from < timedelta(days=2):
            return "hour"
        else:
            return "day"

    @property
    def interval_function(self):
        if self.interval == "day":
            return "toIntervalDay"
        else:
            return "toIntervalHour"


class AppMetricsErrorsQuery(AppMetricsQuery):
    QUERY = QUERY_APP_METRICS_ERRORS
    KEYS = ("error_type", "count", "last_seen")

    def run(self):
        query, params = self.query()
        results = sync_execute(query, params)

        return [dict(zip(self.KEYS, row)) for row in results]


class AppMetricsErrorDetailsQuery:
    QUERY = QUERY_APP_METRICS_ERROR_DETAILS

    def __init__(
        self,
        team: Team,
        plugin_config_id: int,
        filter: AppMetricsErrorsRequestSerializer,
    ):
        self.team = team
        self.plugin_config_id = plugin_config_id
        self.filter = filter

    def run(self):
        query, params = self.query()
        return list(map(self._parse_row, sync_execute(query, params)))

    def query(self):
        job_id = self.filter.validated_data.get("job_id")
        query = self.QUERY.format(job_id_clause="AND job_id = %(job_id)s" if job_id is not None else "")

        return query, {
            "team_id": self.team.pk,
            "plugin_config_id": self.plugin_config_id,
            "category": self.filter.validated_data.get("category"),
            "job_id": job_id,
            "error_type": self.filter.validated_data.get("error_type"),
        }

    def _parse_row(self, row):
        timestamp, error_uuid, error_type, error_details = row
        return {
            "timestamp": timestamp,
            "error_uuid": error_uuid,
            "error_type": error_type,
            "error_details": json.loads(error_details),
        }
