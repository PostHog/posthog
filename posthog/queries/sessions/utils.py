from datetime import datetime
from functools import wraps
from typing import Any, Callable, List, Optional, Tuple, cast

from django.core.cache import cache

from posthog.settings import SESSION_RECORDING_TTL
from posthog.utils import generate_cache_key, get_safe_cache

DistinctId = str
Snapshots = List[Any]


def cached_recording(
    f: Callable[[Any], Tuple[Optional[DistinctId], Optional[datetime], Snapshots]]
) -> Callable[[Any], Tuple[Optional[DistinctId], Optional[datetime], Snapshots]]:
    @wraps(f)
    def wrapper(self) -> Tuple[Optional[DistinctId], Optional[datetime], Snapshots]:
        # Pull from cache if it exists
        cache_key = generate_cache_key("{}_{}".format(self._team.pk, self._session_recording_id))
        cached_data = get_safe_cache(cache_key)

        if cached_data:
            return cached_data

        # Call function being wrapper
        fresh_data = cast(Tuple[Optional[DistinctId], Optional[datetime], Snapshots], f(self))
        # Cache new data
        cache.set(cache_key, fresh_data, SESSION_RECORDING_TTL)
        return fresh_data

    return wrapper
