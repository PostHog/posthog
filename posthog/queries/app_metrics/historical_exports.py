import json
from datetime import timedelta
from typing import Dict, Optional

from zoneinfo import ZoneInfo

from posthog.models.activity_logging.activity_log import ActivityLog
from posthog.models.plugin import PluginStorage
from posthog.models.team.team import Team
from posthog.queries.app_metrics.app_metrics import (
    AppMetricsErrorsQuery,
    AppMetricsQuery,
)
from posthog.queries.app_metrics.serializers import AppMetricsRequestSerializer


def historical_exports_activity(team_id: int, plugin_config_id: int, job_id: Optional[str] = None):
    from posthog.api.shared import UserBasicSerializer

    entries = ActivityLog.objects.filter(
        team_id=team_id,
        scope="PluginConfig",
        item_id=plugin_config_id,
        activity__in=["job_triggered", "export_success", "export_fail"],
        detail__trigger__job_type="Export historical events V2",
        **({"detail__trigger__job_id": job_id} if job_id is not None else {}),
    )

    by_category: Dict = {"job_triggered": {}, "export_success": {}, "export_fail": {}}
    for entry in entries:
        by_category[entry.activity][entry.detail["trigger"]["job_id"]] = entry

    historical_exports = []
    for export_job_id, trigger_entry in by_category["job_triggered"].items():
        record = {
            "created_at": trigger_entry.created_at,
            "created_by": UserBasicSerializer(instance=trigger_entry.user).data,
            "job_id": export_job_id,
            "payload": trigger_entry.detail["trigger"]["payload"],
        }

        if export_job_id in by_category["export_success"]:
            entry = by_category["export_success"][export_job_id]
            record["status"] = "success"
            record["finished_at"] = entry.created_at
            record["duration"] = (entry.created_at - trigger_entry.created_at).total_seconds()
        elif export_job_id in by_category["export_fail"]:
            entry = by_category["export_fail"][export_job_id]
            record["status"] = "fail"
            record["finished_at"] = entry.created_at
            record["duration"] = (entry.created_at - trigger_entry.created_at).total_seconds()
            record["failure_reason"] = entry.detail["trigger"]["payload"].get("failure_reason")
        else:
            record["status"] = "not_finished"
            progress = _fetch_export_progress(plugin_config_id, export_job_id)
            if progress is not None:
                record["progress"] = progress
        historical_exports.append(record)

    historical_exports.sort(key=lambda record: record["created_at"], reverse=True)

    return historical_exports


def historical_export_metrics(team: Team, plugin_config_id: int, job_id: str):
    [export_summary] = historical_exports_activity(team_id=team.pk, plugin_config_id=plugin_config_id, job_id=job_id)

    filter_data = {
        "category": "exportEvents",
        "job_id": job_id,
        "date_from": (export_summary["created_at"] - timedelta(hours=1)).astimezone(ZoneInfo("UTC")).isoformat(),
    }
    if "finished_at" in export_summary:
        filter_data["date_to"] = (
            (export_summary["finished_at"] + timedelta(hours=1)).astimezone(ZoneInfo("UTC")).isoformat()
        )

    filter = AppMetricsRequestSerializer(data=filter_data)
    filter.is_valid(raise_exception=True)
    metric_results = AppMetricsQuery(team, plugin_config_id, filter).run()
    errors = AppMetricsErrorsQuery(team, plugin_config_id, filter).run()

    return {"summary": export_summary, "metrics": metric_results, "errors": errors}


def _fetch_export_progress(plugin_config_id: int, job_id: str) -> Optional[float]:
    coordination_entry = PluginStorage.objects.filter(
        plugin_config_id=plugin_config_id,
        # Keep this in sync with plugin-server/src/worker/vm/upgrades/historical-export/export-historical-events-v2.ts
        key="EXPORT_COORDINATION",
    ).first()

    if coordination_entry is None:
        return None

    return json.loads(coordination_entry.value).get("progress")
