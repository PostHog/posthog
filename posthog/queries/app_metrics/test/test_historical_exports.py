import json
from datetime import datetime
from unittest import mock

from freezegun.api import freeze_time

from posthog.models.activity_logging.activity_log import Detail, Trigger, log_activity
from posthog.models.plugin import Plugin, PluginConfig, PluginStorage
from posthog.models.team.team import Team
from posthog.models.utils import UUIDT
from posthog.queries.app_metrics.historical_exports import (
    historical_export_metrics,
    historical_exports_activity,
)
from posthog.queries.app_metrics.test.test_app_metrics import create_app_metric
from posthog.test.base import (
    BaseTest,
    ClickhouseTestMixin,
    snapshot_clickhouse_queries,
    snapshot_postgres_queries,
)

SAMPLE_PAYLOAD = {"dateRange": ["2021-06-10", "2022-06-12"], "parallelism": 1}


@freeze_time("2021-08-25T13:00:00Z")
class TestHistoricalExports(ClickhouseTestMixin, BaseTest):
    maxDiff = None

    def setUp(self):
        super().setUp()
        self.plugin = Plugin.objects.create(organization=self.organization)
        self.plugin_config = PluginConfig.objects.create(
            pk=3, plugin=self.plugin, team=self.team, enabled=True, order=1
        )

    @snapshot_postgres_queries
    def test_historical_exports_activity_for_not_finished_export(self):
        self._create_activity_log(
            activity="job_triggered",
            detail=Detail(
                name="Some export plugin",
                trigger=Trigger(
                    job_type="Export historical events V2",
                    job_id="1234",
                    payload=SAMPLE_PAYLOAD,
                ),
            ),
        )

        PluginStorage.objects.create(
            plugin_config_id=self.plugin_config.pk,
            key="EXPORT_COORDINATION",
            value=json.dumps({"progress": 0.33}),
        )

        activities = historical_exports_activity(self.team.pk, self.plugin_config.pk)
        self.assertEqual(len(activities), 1)
        self.assertEqual(
            activities[0],
            {
                "job_id": "1234",
                "status": "not_finished",
                "payload": SAMPLE_PAYLOAD,
                "created_at": datetime.fromisoformat("2021-08-25T13:00:00+00:00"),
                "created_by": mock.ANY,
                "progress": 0.33,
            },
        )

    @snapshot_postgres_queries
    def test_historical_exports_activity_for_finished_export(self):
        with freeze_time("2021-08-25T11:00:00Z"):
            self._create_activity_log(
                activity="job_triggered",
                detail=Detail(
                    name="Some export plugin",
                    trigger=Trigger(
                        job_type="Export historical events V2",
                        job_id="1234",
                        payload=SAMPLE_PAYLOAD,
                    ),
                ),
            )
        with freeze_time("2021-08-25T13:00:00Z"):
            self._create_activity_log(
                activity="export_success",
                detail=Detail(
                    name="Some export plugin",
                    trigger=Trigger(
                        job_type="Export historical events V2",
                        job_id="1234",
                        payload={},
                    ),
                ),
            )

        activities = historical_exports_activity(self.team.pk, self.plugin_config.pk)
        self.assertEqual(len(activities), 1)
        self.assertEqual(
            activities[0],
            {
                "job_id": "1234",
                "created_at": datetime.fromisoformat("2021-08-25T11:00:00+00:00"),
                "finished_at": datetime.fromisoformat("2021-08-25T13:00:00+00:00"),
                "status": "success",
                "payload": SAMPLE_PAYLOAD,
                "duration": 2 * 60 * 60,
                "created_by": mock.ANY,
            },
        )

    @snapshot_postgres_queries
    def test_historical_exports_activity_for_failed_export(self):
        with freeze_time("2021-08-25T11:00:00Z"):
            self._create_activity_log(
                activity="job_triggered",
                detail=Detail(
                    name="Some export plugin",
                    trigger=Trigger(
                        job_type="Export historical events V2",
                        job_id="1234",
                        payload=SAMPLE_PAYLOAD,
                    ),
                ),
            )
        with freeze_time("2021-08-25T13:00:00Z"):
            self._create_activity_log(
                activity="export_fail",
                detail=Detail(
                    name="Some export plugin",
                    trigger=Trigger(
                        job_type="Export historical events V2",
                        job_id="1234",
                        payload={"failure_reason": "foobar"},
                    ),
                ),
            )

        activities = historical_exports_activity(self.team.pk, self.plugin_config.pk)
        self.assertEqual(len(activities), 1)
        self.assertEqual(
            activities[0],
            {
                "job_id": "1234",
                "created_at": datetime.fromisoformat("2021-08-25T11:00:00+00:00"),
                "finished_at": datetime.fromisoformat("2021-08-25T13:00:00+00:00"),
                "status": "fail",
                "payload": SAMPLE_PAYLOAD,
                "duration": 2 * 60 * 60,
                "created_by": mock.ANY,
                "failure_reason": "foobar",
            },
        )

    def test_historical_exports_activity_ignores_unrelated_entries(self):
        self._create_activity_log(
            activity="job_triggered",
            detail=Detail(
                name="Some export plugin",
                trigger=Trigger(
                    job_type="Export historical events V2",
                    job_id="1234",
                    payload=SAMPLE_PAYLOAD,
                ),
            ),
        )

        # Different job finishing
        self._create_activity_log(
            activity="export_success",
            detail=Detail(
                name="Some export plugin",
                trigger=Trigger(job_type="Export historical events V2", job_id="another", payload={}),
            ),
        )
        # Different plugin
        self._create_activity_log(
            item_id=2,
            activity="export_success",
            detail=Detail(
                name="Some export plugin",
                trigger=Trigger(job_type="Export historical events V2", job_id="1234", payload={}),
            ),
        )
        # Different plugin
        another_team = Team.objects.create(organization=self.organization)
        self._create_activity_log(
            team_id=another_team.pk,
            activity="export_success",
            detail=Detail(
                name="Some export plugin",
                trigger=Trigger(job_type="Export historical events V2", job_id="1234", payload={}),
            ),
        )
        # Non-export job
        self._create_activity_log(
            team_id=self.team.pk,
            activity="job_triggered",
            detail=Detail(
                name="Some export plugin",
                trigger=Trigger(job_type="Another job", job_id="1234", payload={}),
            ),
        )

        activities = historical_exports_activity(self.team.pk, self.plugin_config.pk)
        self.assertEqual(len(activities), 1)
        self.assertEqual(
            activities[0],
            {
                "job_id": "1234",
                "created_at": datetime.fromisoformat("2021-08-25T13:00:00+00:00"),
                "status": "not_finished",
                "payload": SAMPLE_PAYLOAD,
                "created_by": mock.ANY,
            },
        )

    def test_historical_exports_orders_activity_by_created_at(self):
        for hour in range(10, 15):
            with freeze_time(f"2021-08-25T{hour}:00:00Z"):
                self._create_activity_log(
                    activity="job_triggered",
                    detail=Detail(
                        name="Some export plugin",
                        trigger=Trigger(
                            job_type="Export historical events V2",
                            job_id=str(hour),
                            payload=SAMPLE_PAYLOAD,
                        ),
                    ),
                )

        activities = historical_exports_activity(self.team.pk, self.plugin_config.pk)
        start_times = [activity["created_at"].isoformat() for activity in activities]
        self.assertEqual(
            start_times,
            [
                "2021-08-25T14:00:00+00:00",
                "2021-08-25T13:00:00+00:00",
                "2021-08-25T12:00:00+00:00",
                "2021-08-25T11:00:00+00:00",
                "2021-08-25T10:00:00+00:00",
            ],
        )

    @snapshot_postgres_queries
    @snapshot_clickhouse_queries
    def test_historical_export_metrics(self):
        with freeze_time("2021-08-25T01:00:00Z"):
            self._create_activity_log(
                activity="job_triggered",
                detail=Detail(
                    name="Some export plugin",
                    trigger=Trigger(
                        job_type="Export historical events V2",
                        job_id="1234",
                        payload=SAMPLE_PAYLOAD,
                    ),
                ),
            )
        with freeze_time("2021-08-25T05:00:00Z"):
            self._create_activity_log(
                activity="export_success",
                detail=Detail(
                    name="Some export plugin",
                    trigger=Trigger(
                        job_type="Export historical events V2",
                        job_id="1234",
                        payload={},
                    ),
                ),
            )

        create_app_metric(
            team_id=self.team.pk,
            category="exportEvents",
            plugin_config_id=self.plugin_config.pk,
            job_id="1234",
            timestamp="2021-08-25T01:10:00Z",
            successes=102,
        )
        create_app_metric(
            team_id=self.team.pk,
            category="exportEvents",
            plugin_config_id=self.plugin_config.pk,
            job_id="1234",
            timestamp="2021-08-25T02:55:00Z",
            failures=2,
            error_uuid=str(UUIDT()),
            error_type="SomeError",
            error_details={"event": {}},
        )
        create_app_metric(
            team_id=self.team.pk,
            category="exportEvents",
            plugin_config_id=self.plugin_config.pk,
            job_id="1234",
            timestamp="2021-08-25T03:10:00Z",
            successes=10,
        )

        results = historical_export_metrics(self.team, self.plugin_config.pk, "1234")

        self.assertEqual(
            results,
            {
                "metrics": {
                    "dates": [
                        "2021-08-25 00:00:00",
                        "2021-08-25 01:00:00",
                        "2021-08-25 02:00:00",
                        "2021-08-25 03:00:00",
                        "2021-08-25 04:00:00",
                        "2021-08-25 05:00:00",
                        "2021-08-25 06:00:00",
                    ],
                    "successes": [0, 102, 0, 10, 0, 0, 0],
                    "successes_on_retry": [0, 0, 0, 0, 0, 0, 0],
                    "failures": [0, 0, 2, 0, 0, 0, 0],
                    "totals": {
                        "successes": 112,
                        "successes_on_retry": 0,
                        "failures": 2,
                    },
                },
                "summary": {
                    "duration": 4 * 60 * 60,
                    "finished_at": datetime.fromisoformat("2021-08-25T05:00:00+00:00"),
                    "job_id": "1234",
                    "payload": SAMPLE_PAYLOAD,
                    "created_at": datetime.fromisoformat("2021-08-25T01:00:00+00:00"),
                    "status": "success",
                    "created_by": mock.ANY,
                },
                "errors": [
                    {
                        "error_type": "SomeError",
                        "count": 1,
                        "last_seen": datetime.fromisoformat("2021-08-25T02:55:00+00:00"),
                    },
                ],
            },
        )

    def _create_activity_log(self, **kwargs):
        log_activity(
            **{
                "organization_id": self.team.organization.id,
                "team_id": self.team.pk,
                "user": self.user,
                "item_id": self.plugin_config.pk,
                "scope": "PluginConfig",
                **kwargs,
            }
        )
