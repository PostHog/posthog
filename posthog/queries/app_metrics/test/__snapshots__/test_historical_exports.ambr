# name: TestHistoricalExports.test_historical_export_metrics
  '
  
  SELECT groupArray(date),
         groupArray(successes),
         groupArray(successes_on_retry),
         groupArray(failures)
  FROM
    (SELECT date, sum(successes) AS successes,
                  sum(successes_on_retry) AS successes_on_retry,
                  sum(failures) AS failures
     FROM
       (SELECT dateTrunc('hour', timestamp, 'UTC') AS date,
               sum(successes) AS successes,
               sum(successes_on_retry) AS successes_on_retry,
               sum(failures) AS failures
        FROM app_metrics
        WHERE team_id = 2
          AND plugin_config_id = 3
          AND category = 'exportEvents'
          AND job_id = '1234'
          AND timestamp >= '2021-08-25 00:00:00'
          AND timestamp < '2021-08-25 06:00:00'
        GROUP BY dateTrunc('hour', timestamp, 'UTC'))
     GROUP BY date
     ORDER BY date WITH FILL
     FROM dateTrunc('hour', toDateTime('2021-08-25 00:00:00'), 'UTC') TO dateTrunc('hour', toDateTime('2021-08-25 06:00:00') + toIntervalHour(1), 'UTC') STEP 3600)
  '
---
# name: TestHistoricalExports.test_historical_export_metrics.1
  '
  
  SELECT error_type,
         count() AS count,
         max(timestamp) AS last_seen
  FROM app_metrics
  WHERE team_id = 2
    AND plugin_config_id = 3
    AND category = 'exportEvents'
    AND job_id = '1234'
    AND timestamp >= '2021-08-25 00:00:00'
    AND timestamp < '2021-08-25 06:00:00'
    AND error_type <> ''
  GROUP BY error_type
  ORDER BY count DESC
  '
---
# name: TestHistoricalExports.test_historical_export_metrics.2
  '
  SELECT "posthog_activitylog"."id",
         "posthog_activitylog"."team_id",
         "posthog_activitylog"."organization_id",
         "posthog_activitylog"."user_id",
         "posthog_activitylog"."is_system",
         "posthog_activitylog"."activity",
         "posthog_activitylog"."item_id",
         "posthog_activitylog"."scope",
         "posthog_activitylog"."detail",
         "posthog_activitylog"."created_at"
  FROM "posthog_activitylog"
  WHERE ("posthog_activitylog"."activity" IN ('job_triggered',
                                              'export_success',
                                              'export_fail')
         AND ("posthog_activitylog"."detail" #> ARRAY['trigger',
                                                      'job_id']) = '"1234"'
         AND ("posthog_activitylog"."detail" #> ARRAY['trigger',
                                                      'job_type']) = '"Export historical events V2"'
         AND "posthog_activitylog"."item_id" = '3'
         AND "posthog_activitylog"."scope" = 'PluginConfig'
         AND "posthog_activitylog"."team_id" = 2)
  '
---
# name: TestHistoricalExports.test_historical_export_metrics.3
  '
  SELECT "posthog_user"."id",
         "posthog_user"."password",
         "posthog_user"."last_login",
         "posthog_user"."first_name",
         "posthog_user"."last_name",
         "posthog_user"."is_staff",
         "posthog_user"."is_active",
         "posthog_user"."date_joined",
         "posthog_user"."uuid",
         "posthog_user"."current_organization_id",
         "posthog_user"."current_team_id",
         "posthog_user"."email",
         "posthog_user"."pending_email",
         "posthog_user"."temporary_token",
         "posthog_user"."distinct_id",
         "posthog_user"."is_email_verified",
         "posthog_user"."requested_password_reset_at",
         "posthog_user"."has_seen_product_intro_for",
         "posthog_user"."email_opt_in",
         "posthog_user"."theme_mode",
         "posthog_user"."partial_notification_settings",
         "posthog_user"."anonymize_data",
         "posthog_user"."toolbar_mode",
         "posthog_user"."events_column_config"
  FROM "posthog_user"
  WHERE "posthog_user"."id" = 2
  LIMIT 21
  '
---
# name: TestHistoricalExports.test_historical_exports_activity_for_failed_export
  '
  SELECT "posthog_activitylog"."id",
         "posthog_activitylog"."team_id",
         "posthog_activitylog"."organization_id",
         "posthog_activitylog"."user_id",
         "posthog_activitylog"."is_system",
         "posthog_activitylog"."activity",
         "posthog_activitylog"."item_id",
         "posthog_activitylog"."scope",
         "posthog_activitylog"."detail",
         "posthog_activitylog"."created_at"
  FROM "posthog_activitylog"
  WHERE ("posthog_activitylog"."activity" IN ('job_triggered',
                                              'export_success',
                                              'export_fail')
         AND ("posthog_activitylog"."detail" #> ARRAY['trigger',
                                                      'job_type']) = '"Export historical events V2"'
         AND "posthog_activitylog"."item_id" = '3'
         AND "posthog_activitylog"."scope" = 'PluginConfig'
         AND "posthog_activitylog"."team_id" = 2)
  '
---
# name: TestHistoricalExports.test_historical_exports_activity_for_failed_export.1
  '
  SELECT "posthog_user"."id",
         "posthog_user"."password",
         "posthog_user"."last_login",
         "posthog_user"."first_name",
         "posthog_user"."last_name",
         "posthog_user"."is_staff",
         "posthog_user"."is_active",
         "posthog_user"."date_joined",
         "posthog_user"."uuid",
         "posthog_user"."current_organization_id",
         "posthog_user"."current_team_id",
         "posthog_user"."email",
         "posthog_user"."pending_email",
         "posthog_user"."temporary_token",
         "posthog_user"."distinct_id",
         "posthog_user"."is_email_verified",
         "posthog_user"."requested_password_reset_at",
         "posthog_user"."has_seen_product_intro_for",
         "posthog_user"."email_opt_in",
         "posthog_user"."theme_mode",
         "posthog_user"."partial_notification_settings",
         "posthog_user"."anonymize_data",
         "posthog_user"."toolbar_mode",
         "posthog_user"."events_column_config"
  FROM "posthog_user"
  WHERE "posthog_user"."id" = 2
  LIMIT 21
  '
---
# name: TestHistoricalExports.test_historical_exports_activity_for_finished_export
  '
  SELECT "posthog_activitylog"."id",
         "posthog_activitylog"."team_id",
         "posthog_activitylog"."organization_id",
         "posthog_activitylog"."user_id",
         "posthog_activitylog"."is_system",
         "posthog_activitylog"."activity",
         "posthog_activitylog"."item_id",
         "posthog_activitylog"."scope",
         "posthog_activitylog"."detail",
         "posthog_activitylog"."created_at"
  FROM "posthog_activitylog"
  WHERE ("posthog_activitylog"."activity" IN ('job_triggered',
                                              'export_success',
                                              'export_fail')
         AND ("posthog_activitylog"."detail" #> ARRAY['trigger',
                                                      'job_type']) = '"Export historical events V2"'
         AND "posthog_activitylog"."item_id" = '3'
         AND "posthog_activitylog"."scope" = 'PluginConfig'
         AND "posthog_activitylog"."team_id" = 2)
  '
---
# name: TestHistoricalExports.test_historical_exports_activity_for_finished_export.1
  '
  SELECT "posthog_user"."id",
         "posthog_user"."password",
         "posthog_user"."last_login",
         "posthog_user"."first_name",
         "posthog_user"."last_name",
         "posthog_user"."is_staff",
         "posthog_user"."is_active",
         "posthog_user"."date_joined",
         "posthog_user"."uuid",
         "posthog_user"."current_organization_id",
         "posthog_user"."current_team_id",
         "posthog_user"."email",
         "posthog_user"."pending_email",
         "posthog_user"."temporary_token",
         "posthog_user"."distinct_id",
         "posthog_user"."is_email_verified",
         "posthog_user"."requested_password_reset_at",
         "posthog_user"."has_seen_product_intro_for",
         "posthog_user"."email_opt_in",
         "posthog_user"."theme_mode",
         "posthog_user"."partial_notification_settings",
         "posthog_user"."anonymize_data",
         "posthog_user"."toolbar_mode",
         "posthog_user"."events_column_config"
  FROM "posthog_user"
  WHERE "posthog_user"."id" = 2
  LIMIT 21
  '
---
# name: TestHistoricalExports.test_historical_exports_activity_for_not_finished_export
  '
  SELECT "posthog_activitylog"."id",
         "posthog_activitylog"."team_id",
         "posthog_activitylog"."organization_id",
         "posthog_activitylog"."user_id",
         "posthog_activitylog"."is_system",
         "posthog_activitylog"."activity",
         "posthog_activitylog"."item_id",
         "posthog_activitylog"."scope",
         "posthog_activitylog"."detail",
         "posthog_activitylog"."created_at"
  FROM "posthog_activitylog"
  WHERE ("posthog_activitylog"."activity" IN ('job_triggered',
                                              'export_success',
                                              'export_fail')
         AND ("posthog_activitylog"."detail" #> ARRAY['trigger',
                                                      'job_type']) = '"Export historical events V2"'
         AND "posthog_activitylog"."item_id" = '3'
         AND "posthog_activitylog"."scope" = 'PluginConfig'
         AND "posthog_activitylog"."team_id" = 2)
  '
---
# name: TestHistoricalExports.test_historical_exports_activity_for_not_finished_export.1
  '
  SELECT "posthog_user"."id",
         "posthog_user"."password",
         "posthog_user"."last_login",
         "posthog_user"."first_name",
         "posthog_user"."last_name",
         "posthog_user"."is_staff",
         "posthog_user"."is_active",
         "posthog_user"."date_joined",
         "posthog_user"."uuid",
         "posthog_user"."current_organization_id",
         "posthog_user"."current_team_id",
         "posthog_user"."email",
         "posthog_user"."pending_email",
         "posthog_user"."temporary_token",
         "posthog_user"."distinct_id",
         "posthog_user"."is_email_verified",
         "posthog_user"."requested_password_reset_at",
         "posthog_user"."has_seen_product_intro_for",
         "posthog_user"."email_opt_in",
         "posthog_user"."theme_mode",
         "posthog_user"."partial_notification_settings",
         "posthog_user"."anonymize_data",
         "posthog_user"."toolbar_mode",
         "posthog_user"."events_column_config"
  FROM "posthog_user"
  WHERE "posthog_user"."id" = 2
  LIMIT 21
  '
---
# name: TestHistoricalExports.test_historical_exports_activity_for_not_finished_export.2
  '
  SELECT "posthog_pluginstorage"."id",
         "posthog_pluginstorage"."plugin_config_id",
         "posthog_pluginstorage"."key",
         "posthog_pluginstorage"."value"
  FROM "posthog_pluginstorage"
  WHERE ("posthog_pluginstorage"."key" = 'EXPORT_COORDINATION'
         AND "posthog_pluginstorage"."plugin_config_id" = 2)
  ORDER BY "posthog_pluginstorage"."id" ASC
  LIMIT 1
  '
---
